src/ngx_http_lua_fd.c

```diff
@@ -0,0 +1,221 @@
+        /* queue/fire off handler */
+        r->write_event_handler(r);
+    }
+
+    ngx_http_run_posted_requests(r->connection);
+}
+
+static void ngx_http_lua_fd_cleanup(ngx_http_lua_co_ctx_t *co_ctx) {
+    ngx_http_lua_udata_t *u = co_ctx->data;
+    if (u->conn != NULL) {
+        /* can't use ngx_close_connection here,
+           as it closes the file descriptor unconditionally */
+
+        /* cancel timeout timer */
+        if (u->conn->read->timer_set) {
+            ngx_del_timer(u->conn->read);
+        }
+
+        if (u->conn->fd != -1) {
+            /* remove from mainloop; do not pass CLOSE_SOCKET flag */
+            ngx_del_conn(u->conn, 0);
+        }
+
+        /* delete any pending but not handled events */
+#if (NGX_THREADS)
+        ngx_mutex_lock(ngx_posted_events_mutex);
+#endif
+#if defined(nginx_version) && nginx_version >= 1007005
+        if (u->conn->read->posted) {
+            ngx_delete_posted_event(u->conn->read);
+        }
+        if (u->conn->write->posted) {
+            ngx_delete_posted_event(u->conn->write);
+        }
+#else
+        if (u->conn->read->prev) {
+            ngx_delete_posted_event(u->conn->read);
+        }
+        if (u->conn->write->posted) {
+            ngx_delete_posted_event(u->conn->write);
+        }
+#endif
+#if (NGX_THREADS)
+        ngx_unlock(&u->conn->lock);
+        u->conn->read->locked = 0;
+        u->conn->write->locked = 0;
+
+        ngx_mutex_unlock(ngx_posted_events_mutex);
+#endif
+
+        /* not sure what this line does, the 0 means non-reusable */
+        ngx_reusable_connection(u->conn, 0);
+
+        /* invalidate connection object */
+        u->conn->fd = -1;
+        ngx_free_connection(u->conn);
+
+        u->conn = NULL;
+    }
+    free(u);
+    co_ctx->data = NULL;
+}
+
+static int ngx_http_lua_fd_wait(lua_State *L) {
+    ngx_http_request_t *r;
+    ngx_http_lua_ctx_t *ctx;
+    ngx_http_lua_co_ctx_t *co_ctx;
+    ngx_http_lua_udata_t *u;
+
+    ngx_socket_t fd = luaL_optint(L, 1, -1); /* -1 is invalid fd */
+    const char *events = luaL_optstring(L, 2, "");
+    int poll_mask = 0;
+    double timeout = luaL_optnumber(L, 3, HUGE_VAL); /* default to infinite timeout */
+    while (*events) {
+        if (*events == 'r')
+            poll_mask |= POLLIN;
+        else if (*events == 'w')
+            poll_mask |= POLLOUT;
+        events++;
+    }
+    if ((fd < 0 || !poll_mask) && timeout == HUGE_VAL) {
+        return luaL_error(L, "must provide a valid file descriptor or timeout");
+    }
+    if ((r = ngx_http_lua_get_req(L)) == NULL) {
+        return luaL_error(L, "no request found");
+    }
+    if ((ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module)) == NULL) {
+        return luaL_error(L, "no request ctx found");
+    }
+    if ((co_ctx = ctx->cur_co_ctx) == NULL) {
+        return luaL_error(L, "no co ctx found");
+    }
+    if ((u = calloc(1, sizeof(ngx_http_lua_udata_t))) == NULL) {
```

Comment:
Please avoid using the libc allocator functions directly. Use the nginx wrappers instead.